File: typescript/packages/core/src/commands/cli/builtin/git/log.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult } from '../../../../io/types.ts'
import type { CommandFnResult } from '../../../config.ts'
import { FlagView } from '../../../spec/flag_view.ts'
import type { CLIInvocation } from '../../types.ts'
import { GitError } from './errors.ts'
import {
  joinOutput,
  commitOutput,
  parseDiffFlags,
  renamesEnabled,
  separatorLine,
  type DiffFlags,
} from './diff_output.ts'
import {
  FULL_SHA,
  needsDecorations,
  oneline,
  presetBlock,
  renderTemplate,
  type CommitFacts,
  type Decorations,
} from './format.ts'
import { CommitGraph } from './graph.ts'
import {
  decorations,
  parseFlags,
  refCommits,
  select,
  walked,
  type LogFlags,
  type Walk,
} from './history.ts'
import { configBool, opened, type Repo } from './repo.ts'
import { splitRevisions } from './revparse.ts'
import { checkOperands, escaped, fatal } from './util.ts'
import { HEAD } from './constants.ts'
import { encodeText } from '../../../../shell/bytes.ts'

/**
 * The text a log invocation prints for its selected commits.
 *
 * `format:` separates entries with a newline and ends without one, and an
 * entry that renders empty still claims its separator, so `--pretty=format:`
 * prints one newline per commit past the first. `tformat:` (and any bare `%`
 * string) terminates every entry, empty ones included - except that an empty
 * template prints nothing at all, which is how `--format=` stays silent.
 * The caller encodes through `encodeText` because `%xHH` names a raw byte.
 * Pinned against git 2.37 and 2.54.
 */
function rendered(
  commits: readonly CommitFacts[],
  flags: LogFlags,
  width: number,
  decor: Decorations | null,
): string {
  const fmt = flags.pretty
  if (fmt.kind === 'oneline') {
    const length = flags.abbrevCommit ? width : FULL_SHA
    const lines = commits.map((commit) =>
      flags.decorate ? renderTemplate('%h%d %s', commit, length, decor) : oneline(commit, length),
    )
    return lines.length > 0 ? `${lines.join('\n')}\n` : ''
  }
  if (fmt.kind === 'format' || fmt.kind === 'tformat') {
    const entries = commits.map((commit) =>
      renderTemplate(fmt.template ?? '', commit, width, decor, flags.date),
    )
    if (fmt.kind === 'tformat') {
      if (fmt.template === null || fmt.template === '') return ''
      return entries.map((text) => `${text}\n`).join('')
    }
    return entries.join('\n')
  }
  const lines: string[] = []
  commits.forEach((commit, index) => {
    if (index > 0) lines.push('')
    const block = presetBlock(commit, fmt.kind, width, flags.date)
    if (flags.decorate && block[0]?.startsWith('commit '))
      block[0] += renderTemplate('%d', commit, width, decor)
    lines.push(...block)
  })
  return lines.length > 0 ? `${lines.join('\n')}\n` : ''
}

/**
 * The text a `--graph` log prints: git's show_log, commit by commit.
 *
 * Every walked commit moves the graph on, printed or not, so a commit the
 * pickaxe passed by leaves a `...` row. A printed commit gets the graph's lines
 * up to its own, then its header, then its text with the next graph line in
 * front of each further line, then whatever lines the graph still owes. The
 * formats that separate entries (medium and its kin, `format:`) put the
 * separator behind a padding line, so the graph never shows a gap; the ones
 * that terminate entries (oneline, `tformat:`) do the same after each entry.
 * A padding line is skipped wherever the text before it ended without a
 * newline, since it would then land on that text's own line.
 *
 * With a diff each block (one per parent under `-m`) is an entry of its own,
 * each naming its parent, and every diff line sits behind a padding line,
 * the one between the message and the diff included. That line is `---` when
 * both a diffstat and a patch follow, and is left out for oneline, except
 * before a combined diff, which git prints from its own path.
 *
 * @param repo the opened repository
 * @param walk the walked commits and the ones an edge may lead to
 * @param flags the parsed invocation
 * @param decor ref labels per commit, when the format prints any
 * @param diff the diff flags, null when no diff was asked for
 */
async function graphed(
  repo: Repo,
  walk: Walk,
  flags: LogFlags,
  decor: Decorations | null,
  diff: DiffFlags | null,
): Promise<string> {
  const width = repo.abbrev
  const graph = new CommitGraph((oid) => walk.interesting.has(oid), flags.firstParent)
  const fmt = flags.pretty
  const user = fmt.kind === 'format' || fmt.kind === 'tformat'
  const terminated = fmt.kind === 'oneline' || fmt.kind === 'tformat'
  const empty = user && (fmt.template ?? '') === ''
  const length = fmt.kind === 'oneline' && !flags.abbrevCommit ? FULL_SHA : width
  let out = ''
  let shownOne = false
  let missingNewline = false
  for (const { commit, shown } of walk.steps) {
    graph.update(commit)
    if (!shown) continue
    const bodies = diff === null ? [] : await commitOutput(repo, commit, diff)
    const blocks = bodies.length === 0 ? [''] : bodies
    blocks.forEach((body, index) => {
      if (shownOne && !terminated) {
        if (!missingNewline) out += graph.paddingLine()
        out += '\n'
      }
      shownOne = true
      out += graph.showCommit()
      const parent = commit.parents[index]
      const from =
        bodies.length > 1 && !user && parent !== undefined
          ? ` (from ${parent.slice(0, fmt.kind === 'oneline' ? length : FULL_SHA)})`
          : ''
      const labels = flags.decorate ? renderTemplate('%d', commit, width, decor) : ''
      let text: string
      if (fmt.kind === 'oneline') {
        out += `${renderTemplate('%h', commit, length, decor)}${from}${labels} `
        text = renderTemplate('%s', commit, length, decor)
      } else if (user) {
        text = renderTemplate(fmt.template ?? '', commit, width, decor, flags.date)
      } else {
        const [head = '', ...rest] = presetBlock(commit, fmt.kind, width, flags.date)
        out += `${head}${from}${labels}\n${graph.nextLine()[0]}`
        text = rest.map((line) => `${line}\n`).join('')
      }
      missingNewline = !text.endsWith('\n')
      out += graph.showMessage(text)
      if (terminated && !empty) {
        if (!missingNewline) out += graph.paddingLine()
        out += '\n'
      }
      if (body === '' || diff === null) return
      const separator = empty ? null : separatorLine(commit, fmt.kind, diff)
      if (separator !== null) out += `${graph.paddingLine()}${separator}\n`
      for (const line of body.split('\n').slice(0, -1)) out += `${graph.paddingLine()}${line}\n`
    })
  }
  return out
}

/**
 * The commits a log walks from and the commits it hides: the revisions and
 * ranges given, HEAD when there are none, plus every ref for --all.
 */
async function startingPoints(
  repo: Repo,
  revisions: readonly string[],
  flags: LogFlags,
): Promise<[CommitFacts[], CommitFacts[]]> {
  const [starts, hidden] = await splitRevisions(repo, revisions.length ? revisions : [HEAD])
  if (flags.allRefs) starts.push(...(await refCommits(repo)))
  return [starts, hidden]
}

/** Show commit logs. */
export async function log(inv: CLIInvocation): Promise<CommandFnResult> {
  const doors = inv.doors ?? {}
  const texts = [...inv.texts]
  const fl = new FlagView(inv.flags)
  try {
    checkOperands(texts, undefined, escaped(inv.argv))
    const parsed = parseFlags(fl)
    const repo = await opened(fl, doors)
    const [starts, hidden] = await startingPoints(repo, texts, parsed)
    const decor =
      parsed.decorate || needsDecorations(parsed.pretty) ? await decorations(repo) : null
    let diffFlags = parseDiffFlags(fl, false)
    const diffing =
      diffFlags.patch ||
      diffFlags.stat ||
      diffFlags.nameOnly ||
      diffFlags.nameStatus ||
      diffFlags.numstat ||
      diffFlags.shortstat ||
      diffFlags.summary ||
      diffFlags.raw
    if (diffing) {
      diffFlags = parseDiffFlags(
        fl,
        false,
        'off',
        true,
        await renamesEnabled(repo),
        await configBool(repo, 'core.quotepath', true),
      )
    }
    if (parsed.graph) {
      const walk = await walked(repo, starts, parsed, hidden)
      const out = await graphed(repo, walk, parsed, decor, diffing ? diffFlags : null)
      return [out === '' ? null : encodeText(out), new IOResult()]
    }
    const commits = await select(repo, starts, parsed, hidden)
    let out: string
    if (diffing) {
      const blocks: string[] = []
      for (const commit of commits) {
        const head = rendered([commit], parsed, repo.abbrev, decor)
        const bodies = await commitOutput(repo, commit, diffFlags)
        blocks.push(joinOutput(commit, head, bodies, parsed.pretty.kind, repo.abbrev, diffFlags))
      }
      out = blocks.join(['tformat', 'oneline'].includes(parsed.pretty.kind) ? '' : '\n')
    } else out = rendered(commits, parsed, repo.abbrev, decor)
    if (out === '') return [null, new IOResult()]
    return [encodeText(out), new IOResult()]
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
}

File: python/mirage/commands/cli/builtin/git/log.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio

from dulwich.objects import Commit
from dulwich.repo import BaseRepo

from mirage.commands.cli.builtin.git.constants import HEAD
from mirage.commands.cli.builtin.git.diff_output import (  # yapf: disable
    DiffFlags, commit_output, join_output, parse_diff_flags, renames_enabled,
    separator_line)
from mirage.commands.cli.builtin.git.errors import GitError, NoWorkspaceError
from mirage.commands.cli.builtin.git.format import (FULL_SHA, Decorations,
                                                    needs_decorations, oneline,
                                                    preset_block,
                                                    render_template)
from mirage.commands.cli.builtin.git.graph import CommitGraph
from mirage.commands.cli.builtin.git.history import (LogFlags, Walk,
                                                     decorations, parse_flags,
                                                     ref_commits, select,
                                                     walked)
from mirage.commands.cli.builtin.git.objects import abbrev_for
from mirage.commands.cli.builtin.git.repo import config_bool
from mirage.commands.cli.builtin.git.revparse import split_revisions
from mirage.commands.cli.builtin.git.session import opened
from mirage.commands.cli.builtin.git.util import (  # yapf: disable
    check_operands, escaped, fatal)
from mirage.commands.cli.types import CLIDoors, CLIInvocation
from mirage.commands.spec.flag_view import FlagView
from mirage.io.stream import yield_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.shell.bytes import encode_text


def _collect(
        repo: BaseRepo, revisions: tuple[str, ...], flags: LogFlags,
        want_decor: bool
) -> tuple[list[Commit], Walk | None, Decorations | None]:
    """Resolve the starting points and walk them, synchronously.

    Runs on a worker thread. dulwich's walker is synchronous and now
    fetches objects as it goes, so it has to sit off the event loop that
    serves those fetches; enumerating refs and peeling tags fetch the
    same way, which is why ``--all`` and the decoration table are
    resolved here too.

    Args:
        repo (BaseRepo): repository to walk.
        revisions (tuple[str, ...]): the revisions and ranges to walk,
            HEAD when none was given.
        flags (LogFlags): the parsed invocation.
        want_decor (bool): whether the format renders %d/%D.
    """
    starts, hidden = split_revisions(repo, revisions or (HEAD, ))
    if flags.all_refs:
        starts.extend(ref_commits(repo))
    decor = decorations(repo) if want_decor else None
    if flags.graph:
        return [], walked(repo, starts, flags, tuple(hidden)), decor
    return select(repo, starts, flags, tuple(hidden)), None, decor


def _rendered(commits: list[Commit], flags: LogFlags, width: int,
              decor: Decorations | None) -> bytes:
    """The bytes a log invocation prints for its selected commits.

    ``format:`` separates entries with a newline and ends without one,
    and an entry that renders empty still claims its separator, so
    ``--pretty=format:`` prints one newline per commit past the first.
    ``tformat:`` (and any bare ``%`` string) terminates every entry,
    empty ones included - except that an empty template prints nothing
    at all, which is how ``--format=`` stays silent. Bytes go out
    through ``encode_text`` because ``%xHH`` names a raw byte. Pinned
    against git 2.37 and 2.54.

    Args:
        commits (list[Commit]): the selected commits, in print order.
        flags (LogFlags): the parsed invocation.
        width (int): abbreviated id width for this repository.
        decor (Decorations | None): ref labels when the format asked.
    """
    fmt = flags.pretty
    if fmt.kind == "oneline":
        length = width if flags.abbrev_commit else FULL_SHA
        lines = [
            render_template("%h%d %s", commit, length, decor)
            if flags.decorate else oneline(commit, length)
            for commit in commits
        ]
        return ("\n".join(lines) + "\n").encode() if lines else b""
    if fmt.kind in ("format", "tformat"):
        rendered = [
            render_template(fmt.template or "", commit, width, decor,
                            flags.date) for commit in commits
        ]
        if fmt.kind == "tformat":
            if not fmt.template:
                return b""
            return encode_text("".join(f"{text}\n" for text in rendered))
        return encode_text("\n".join(rendered))
    lines = []
    for index, commit in enumerate(commits):
        if index:
            lines.append("")
        block = preset_block(commit, fmt.kind, width, flags.date)
        if flags.decorate and block and block[0].startswith("commit "):
            block[0] += render_template("%d", commit, width, decor)
        lines.extend(block)
    return ("\n".join(lines) + "\n").encode() if lines else b""


def _graphed(repo: BaseRepo, walk: Walk, flags: LogFlags,
             decor: Decorations | None, diff: DiffFlags | None) -> bytes:
    """The bytes a ``--graph`` log prints: git's show_log, commit by
    commit.

    Every walked commit moves the graph on, printed or not, so a commit
    the pickaxe passed by leaves a ``...`` row. A printed commit gets the
    graph's lines up to its own, then its header, then its text with the
    next graph line in front of each further line, then whatever lines
    the graph still owes. The formats that separate entries (medium and
    its kin, ``format:``) put the separator behind a padding line, so
    the graph never shows a gap; the ones that terminate entries
    (oneline, ``tformat:``) do the same after each entry. A padding line
    is skipped wherever the text before it ended without a newline,
    since it would then land on that text's own line.

    With a diff each block (one per parent under ``-m``) is an entry of
    its own, each naming its parent, and every diff line sits behind a
    padding line, the one between the message and the diff included.
    That line is ``---`` when both a diffstat and a patch follow, and is
    left out for oneline, except before a combined diff, which git
    prints from its own path. Synchronous, for a worker thread: the
    diffs read objects.

    Args:
        repo (BaseRepo): the opened repository.
        walk (Walk): the walked commits and the ones an edge may lead
            to.
        flags (LogFlags): the parsed invocation.
        decor (Decorations | None): ref labels per commit, when the
            format prints any.
        diff (DiffFlags | None): the diff flags, None when no diff was
            asked for.
    """
    width = abbrev_for(repo)
    graph = CommitGraph(walk.interesting.__contains__, flags.first_parent)
    fmt = flags.pretty
    user = fmt.kind in ("format", "tformat")
    terminated = fmt.kind in ("oneline", "tformat")
    empty = user and not fmt.template
    length = FULL_SHA if fmt.kind == "oneline" and not flags.abbrev_commit \
        else width
    out = ""
    shown_one = False
    missing_newline = False
    for step in walk.steps:
        commit = step.commit
        graph.update(commit)
        if not step.shown:
            continue
        bodies = [
            body.decode("utf-8", "surrogateescape")
            for body in commit_output(repo, commit, diff)
        ] if diff is not None else []
        for index, body in enumerate(bodies or [""]):
            if shown_one and not terminated:
                if not missing_newline:
                    out += graph.padding_line()
                out += "\n"
            shown_one = True
            out += graph.show_commit()
            parent = (commit.parents[index].decode()
                      if index < len(commit.parents) else None)
            source = ""
            if len(bodies) > 1 and not user and parent is not None:
                cut = length if fmt.kind == "oneline" else FULL_SHA
                source = f" (from {parent[:cut]})"
            labels = (render_template("%d", commit, width, decor)
                      if flags.decorate else "")
            if fmt.kind == "oneline":
                out += (f"{render_template('%h', commit, length, decor)}"
                        f"{source}{labels} ")
                text = render_template("%s", commit, length, decor)
            elif user:
                text = render_template(fmt.template or "", commit, width,
                                       decor, flags.date)
            else:
                head, *rest = preset_block(commit, fmt.kind, width, flags.date)
                out += f"{head}{source}{labels}\n{graph.next_line()[0]}"
                text = "".join(f"{line}\n" for line in rest)
            missing_newline = not text.endswith("\n")
            out += graph.show_message(text)
            if terminated and not empty:
                if not missing_newline:
                    out += graph.padding_line()
                out += "\n"
            if body == "" or diff is None:
                continue
            separator = None if empty else separator_line(
                commit, fmt.kind, diff)
            if separator is not None:
                out += f"{graph.padding_line()}{separator}\n"
            for line in body.split("\n")[:-1]:
                out += f"{graph.padding_line()}{line}\n"
    return encode_text(out)


async def log(inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    """Show commit logs.

    Args:
        inv (CLIInvocation[None]): the line's invocation record.
            git declares no config_model; the planes it reads
            (data through ``dispatch``, names through ``ns``) ride
            ``inv.doors``.
    """
    doors = inv.doors or CLIDoors()
    dispatch = doors.dispatch
    texts = inv.texts
    flags = inv.flags
    fl = FlagView(flags)
    try:
        if dispatch is None:
            raise NoWorkspaceError()
        check_operands(texts, marked=escaped(inv.argv))
        parsed = parse_flags(fl)
        repo, _location = await opened(fl, doors)
        commits, walk, decor = await asyncio.to_thread(
            _collect, repo, tuple(texts), parsed,
            (parsed.decorate or needs_decorations(parsed.pretty)))
        diff_flags = parse_diff_flags(fl, default_patch=False)
    except GitError as exc:
        return fatal(exc)
    diffing = any((diff_flags.patch, diff_flags.stat, diff_flags.name_only,
                   diff_flags.name_status, diff_flags.numstat,
                   diff_flags.shortstat, diff_flags.summary, diff_flags.raw))
    if diffing:
        diff_flags = parse_diff_flags(fl,
                                      default_patch=False,
                                      default_renames=await
                                      renames_enabled(dispatch, _location),
                                      quote_path_fully=await
                                      config_bool(dispatch, _location, b"core",
                                                  b"quotepath", True))
    if walk is not None:
        out = await asyncio.to_thread(_graphed, repo, walk, parsed, decor,
                                      diff_flags if diffing else None)
    elif diffing:
        blocks = []
        for commit in commits:
            head = _rendered([commit], parsed, abbrev_for(repo), decor)
            bodies = await asyncio.to_thread(commit_output, repo, commit,
                                             diff_flags)
            blocks.append(
                join_output(commit, head, bodies, parsed.pretty.kind,
                            abbrev_for(repo), diff_flags))
        out = (b"" if parsed.pretty.kind in ("tformat", "oneline") else
               b"\n").join(blocks)
    else:
        out = _rendered(commits, parsed, abbrev_for(repo), decor)
    if not out:
        return None, IOResult()
    return yield_bytes(out), IOResult()
